### DISCUSSION TITLE
Cameras on 3g / lower bandwidth connections

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#595] [Entity Models]
When using Picture entity card, camera images are returned from the backend without being scaled as it relies on the browser to scale the image.  For 4k cameras, this means megabyte or larger images being loaded every few seconds.  Keeping the Home Assistant app open quickly consumes 100s of megabytes of data with just a few cameras.  

On slower connections, or when the home internet provider doesn't have a fast upload rate (3 mbps is considered broadband in the US - https://www.theverge.com/2015/1/29/7932653/fcc-changed-definition-broadband-25mbps), the user experience is broken camera images.
<img width="397" alt="Screen Shot 2021-07-31 at 2 59 41 PM" src="https://user-images.githubusercontent.com/663432/127751164-c43394e3-08fb-491b-a4fd-0687eca73b53.png">

The visible image usually only requires ~15-70k when scaled since the user is usually only viewing at most a 480 px height card.

Proposal for the following:
- Add a `width` and `height` argument to async_camera_image
  
  If the camera supports returning a scaled image, the width and height are passed to the camera and the overhead of scaling the image is offloaded to the camera.  This is the ideal scenario since the camera will likely have hardware acceleration to do this if the api supports it

  If the camera does not have an api, scale the image ourselves on the backend using the same method we already do in HomeKit: https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit/img_util.py

**Comment by DubhAd** [comment]
I know various cameras also offer lower resolution streams. Using those directly or even downscaling those should be a lot lower impact than downscaling a 4k (or higher) image.

Clearly, that only works where the camera supports it, and would require that the camera platform supported those too, but it's worth a thought.

**Reply by bdraco** [reply]
I *think* as long as we pass on the width and height to the integration it should have enough information be able to make the decision on where to get the snapshot from (camera api, ffmpeg snapshot of the stream, snapshot of the lower resolution stream, etc).

**Reply by DubhAd** [reply]
But, surely it needs to know about the low res stream? Do all the camera integrations support that already?

**Reply by Jc2k** [reply]
If the integration knows what size we are after it can make the right choice based on the streams it knows about. For example some integrations don't even need to use a stream at all because the camera has an api to get a single frame at the target resolution. But that's up to the individual integrations to care about.

**Reply by DubhAd** [reply]
Which means ideally integrations need extended to be aware of lower resolution streams/snapshots IMO

**Reply by Jc2k** [reply]
Yes an integration could be extended to implement this API using a more efficient low res stream. And as I said others can just use their respective stills APIs without touching any streams at all. It's up to the integrations to implement this API the most efficient way they can.

I don't think it makes sense to talk about how individual integrations implement it though? We just need to agree an API for the integration to implement here? Or is there a problem with the API?

As for accessing the actual stream at a lower res, that's probably a different arch issue and we shouldn't hijack this one to talk about it?

**Reply by thecode** [reply]
I am also wondering what would be the right approach, maybe we should supply a value for the maximum resolution, so that the integration can make a better decision. The camera I use for example has a sub-stream of 640x360. What would be the required outcome?
Let's assume the following cases:
1. Request height is 480px which is higher than the camera sub-steam and lower than  the native resolution - 1080px.
2. Requested height is 320px, lower than camera sub-stream.

In both cases I think camera or backend scaling should be avoided.

**Reply by bdraco** [reply]
The only input data we are going to have in all cases (Home Assistant UI, HomeKit, etc) is `width` and `height`.  If we start making guesses on maximum resolution we will probably get it wrong in some cases.

I think it is best to leave it up to the integration to make the decision based on the `width` and `height` passed.  Scaling should be done on a best effort basis, not a requirement as we can always fallback to having the display layer scale it if needed which is what we have now.

**Comment by bdraco** [comment]
PR that implements this
https://github.com/home-assistant/core/pull/53835

**Comment by bdraco** [comment]
In a followup to this, I plan on integrating this with the existing proxy `max_image_width`/`max_image_height` https://www.home-assistant.io/integrations/proxy/#max_image_width to avoid the overhead of scaling if the image is already scaled by the camera / device as it uses `PIL.Image` which is roughly an order of magnitude slower (why its not used in img_util).

**Comment by balloob** [comment]
WebRTC has automatic bandwidth adjustments. Is this still needed?
